revpi.py

```diff
@@ -1,89 +1,141 @@
+        ret = fcntl.ioctl(self.f, prm, struct_name)
+        return ret
+
+    def write(self, name, value):
+        """
+        Set the value of a variable (most likely an output)
+        :param name: name of the variable as defined in the Pictory
+        :param value: value to be set
+        """
+        if name not in self.positions:
+            self.positions[name] = self.get_position(name)
+        offset = struct.unpack_from('>H', self.positions[name], 32)[0]
+        length = struct.unpack_from('>H', self.positions[name], 35)[0]
+        prm = (b'K'[0] << 8) + 16
+        byte_array = bytearray([0, 0, 0, 0])
+        if length == 1:  # single bit
+            if value in [0, 1]:
+                bit = struct.unpack_from('B', self.positions[name], 34)[0]
+                struct.pack_into('>H', byte_array, 0, offset)
+                struct.pack_into('B', byte_array, 2, bit)
+                struct.pack_into('B', byte_array, 3, value)
+                fcntl.ioctl(self.f, prm, byte_array)
+            else:
+                self.logger.debug(f'Invalid value {value}. choose 0 or 1')
+        else:  # writing 2 bytes
+            self.f.seek(int(offset >> 8))
+            self.f.write(int(value).to_bytes(2, 'little'))
+
+    def read(self, name):
+        """
+        Read value of a variable
+        :param name: name of the variable as defined in the Pictory
+        """
+        if name not in self.positions:
+            self.positions[name] = self.get_position(name)
+        value = bytearray([0, 0, 0, 0])
+        offset = struct.unpack_from('>H', self.positions[name], 32)[0]
+        length = struct.unpack_from('>H', self.positions[name], 35)[0]
+        prm = (b'K'[0] << 8) + 15
+        if length == 1:  # single bit
+            bit = struct.unpack_from('B', self.positions[name], 34)[0]
+            struct.pack_into('>H', value, 0, offset)
+            struct.pack_into('B', value, 2, bit)
+            fcntl.ioctl(self.f, prm, value)
+            ret = value[3]
+        else:  # two bytes
+            self.f.seek(int(offset >> 8))
+            ret = int.from_bytes(self.f.read(2), 'little')
+        return ret
+
     def execute_command(self, command):
         # command looks like "set <target> <value>"
         # this is wrapped in a try-except one call up so we don't
         # need to do it here. We rsplit with max here so "target"
         # can potentially have spaces in it and won't cause issues
         target, value = command[4:].rsplit(' ', maxsplit=1)
-        if target == 'fast_cooling_valve':
+        if target in self.targets:
             # "fast_cooling_valve open" or something
-            module = 3
-            ch = 13
-            value = 1 if value == 'open' else 0
-        return self.commands['writeOutput'].format(module=module, ch=output, value=value)
+            target = self.targets[target]
+            if value in self.keywords:
+                value = self.keywords[value]
```

Comment:
We could compact these lines with something like `target=self.targets.get(target, target)` and `value=self.keywords.get(value, value)`

Reply:
Done